rfd/0171-database-session-playback.md

```diff
@@ -0,0 +1,376 @@
+following values:
+- `on`: Enables session recording, but only queries are recorded. This will
+  mimic the current behavior.
+- `full`: Enables session recording. Queries and responses are recorded.
+- `off`: Disables session recording. Audit events are kept unchanged (start,
+  end, and query events emitted).
+- `recording_only`: Enables recording, but events will not be emitted, they
+  will only be present on recordings. This mode will be the same format as SSH
+  sessions, where the commands and their results are only present on the
+  recording.
+
+The value will be set to `on` by default, keeping the current behavior.
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: alice
+spec:
+  allow:
+    ...
+  options:
+    record_session:
+      db: on|full|recording_only|off
+```
+
+### Player
+
+To play the session recording, we will rely on the existing player used by SSH
+sessions. This means we don't need to introduce any new player on `tsh` or the
+Web UI. We're going to convert database recording events into `SessionPrint`
+events, which can be rendered by the players.
+
+This conversion will be agnostic to the database protocol. This will make it
+easier to keep the session recording play consistent across protocols.
+
+It will present the session recording events in different text formats:
+
+- User queries will be presented as received with additional information about
+  the database name.
+- Data with fields metadata: Format the data into an ASCII table, where each
+  field is a column.
+- Data without field metadata: Print each row as a single line.
+- Status (result) without data: Show the user message. If none is present, show
+  "OK" in case of success or "ERROR" in case of error.
+
+Example of player visualization:
+
+```code
+mydatabse=# SELECT id, name FROM events;
+| id |      name     |
++----+---------------+
+|  1 | session.start |
+|  2 | session.end   |
+
+mydatabase=# INSERT INTO events (name) VALUES ('session.query');
+
+1 row affected
+
+mydatabase=# SELECT with_error;
+ERROR: column "with_error" does not exist
+
+mydatabase=# ALTER SYSTEM SET client_min_messages = 'notice';
+OK
+
+(session end)
+```
+
+### Performance considerations
+
+Recording queries and responses will increase storage requirements and
+potentially impact CPU and memory usage. To reduce these, the following measures
+will be taken:
+
+- The recording mode will be similar to `node` (async), where the recording is
+  always done on the node side, persisted into the local filesystem, and
+  uploaded to the auth server after the session ends.
+- Server response events will be generated and recorded in a separate goroutine,
+  avoiding delayed message delivery to the clients.
+- Record a predefined maximum number of rows per result. If this limit is
+  exceeded, the rows will be truncated. In addition to this limit, the record
+  will also ensure the Protobuf max message size is not exceeded.
```

Comment:
I think it would be useful to expose those limits to the user; see the proposed configuration schema above.